### PR TITLE
router: Add deferred reset error details

### PR DESCRIPTION
**Pull request by abeyad**

### CODE REVIEW

File: source/common/router/upstream_codec_filter.cc [hunk 1]
```diff
@@ -69,9 +69,10 @@ Http::FilterHeadersStatus UpstreamCodecFilter::decodeHeaders(Http::RequestHeader
     // It is possible that encodeHeaders() fails. This can happen if filters or other extensions
     // erroneously remove required headers.
     callbacks_->streamInfo().setResponseFlag(StreamInfo::CoreResponseFlag::DownstreamProtocolError);
-    const std::string details =
-        absl::StrCat(StreamInfo::ResponseCodeDetails::get().FilterRemovedRequiredRequestHeaders,
-                     "{", StringUtil::replaceAllEmptySpace(status.message()), "}");
+    const std::string details = absl::StrCat(
+        StreamInfo::ResponseCodeDetails::get().FilterRemovedRequiredRequestHeaders, "{",
+        deferred_reset_ ? "deferred_reset" : StringUtil::replaceAllEmptySpace(status.message()),
```
Comment (alyssawilk) [inline source/common/router/upstream_codec_filter.cc:74]: nice catch
this is Ok, but WDYT about making defered_reset_ instead an optional status defered_reset_status_ so we get the original details?
Also please test even if it's a unit test.
/wait

Comment (abeyad) [reply source/common/router/upstream_codec_filter.cc]: Thanks for the suggestions, addressed your feedback, PTAL!

**Review by alyssawilk — DISMISSED** [review]
Very nice!

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**